x-pack/legacy/plugins/maps/public/layers/blended_vector_layer.ts

```diff
@@ -0,0 +1,227 @@
+      label: COUNT_PROP_LABEL,
+    },
+    ...documentStyle.getDynamicPropertiesArray().map(dynamicProperty => {
+      return {
+        type: getAggType(dynamicProperty),
+        field: dynamicProperty.getFieldName(),
+      };
+    }),
+  ];
+  return new ESGeoGridSource(clusterSourceDescriptor, documentSource.getInspectorAdapters());
+}
+
+function getClusterStyleDescriptor(
+  documentStyle: IVectorStyle,
+  clusterSource: IESAggSource
+): unknown {
+  const defaultDynamicProperties = getDefaultDynamicProperties();
+  const clusterStyleDescriptor: any = {
+    ...documentStyle.getDescriptor(),
+    properties: {
+      [VECTOR_STYLES.LABEL_TEXT]: {
+        type: STYLE_TYPE.DYNAMIC,
+        options: {
+          ...defaultDynamicProperties[VECTOR_STYLES.LABEL_TEXT].options,
+          field: {
+            name: COUNT_PROP_NAME,
+            origin: SOURCE_DATA_ID_ORIGIN,
+          },
+        },
+      },
+      [VECTOR_STYLES.ICON_SIZE]: {
+        type: STYLE_TYPE.DYNAMIC,
+        options: {
+          ...defaultDynamicProperties[VECTOR_STYLES.ICON_SIZE].options,
+          field: {
+            name: COUNT_PROP_NAME,
+            origin: SOURCE_DATA_ID_ORIGIN,
+          },
+        },
+      },
+    },
+  };
+  documentStyle.getAllStyleProperties().forEach((styleProperty: IStyleProperty) => {
+    const styleName = styleProperty.getStyleName();
+    if (
+      [VECTOR_STYLES.LABEL_TEXT, VECTOR_STYLES.ICON_SIZE].includes(styleName) &&
+      (!styleProperty.isDynamic() || !styleProperty.isComplete())
+    ) {
+      // Do not migrate static label and icon size properties to provide unique cluster styling out of the box
+      return;
+    }
+
+    if (styleProperty.isDynamic()) {
+      const options = (styleProperty as IDynamicStyleProperty).getOptions();
+      const field =
+        options && options.field && options.field.name
+          ? {
+              ...options.field,
+              name: clusterSource.getAggKey(
+                getAggType(styleProperty as IDynamicStyleProperty),
+                options.field.name
+              ),
+            }
+          : undefined;
+      clusterStyleDescriptor.properties[styleName] = {
+        type: STYLE_TYPE.DYNAMIC,
+        options: {
+          ...options,
+          field,
+        },
+      };
+    } else {
+      clusterStyleDescriptor.properties[styleName] = {
+        type: STYLE_TYPE.STATIC,
+        options: { ...styleProperty.getOptions() },
+      };
+    }
+  });
+
+  return clusterStyleDescriptor;
+}
+
+export class BlendedVectorLayer extends VectorLayer implements IVectorLayer {
+  static type = LAYER_TYPE.BLENDED_VECTOR;
+
+  static createDescriptor(options: VectorLayerArguments, mapColors: string[]) {
+    const layerDescriptor = VectorLayer.createDescriptor(options, mapColors);
+    layerDescriptor.type = BlendedVectorLayer.type;
+    return layerDescriptor;
+  }
+
+  private _activeSource: IESSource;
+  private _activeStyle: IVectorStyle;
```

Comment:
that these are _not_ readonly is exceptional. Virtually all non React-component instances (basically, the data model under `maps/public/layer/**`) in Maps are frozen. This is the only component where this is not the case. 

the main reason for mutable state here is that we want to "pass" some knowledge from `syncData` to `syncLayerWithMb`. (ie. activeSource and activeStyle).

I'm not sure if using private-state is the right way to do this, as it breaks the immutability of our instances. It also implicitly relies on the fact that our instances don't get clobbered in between our actions calling `Layer#syncData` and the mapbox control syncing style with `Layer#syncLayerWithMb`

I'm wondering if it's worth doing this differently.

My high-level proposal would be:
- to determine active-style by evalutating the `getMaxResultWindow > maxResultWindow` in `getCurrentStyle`. `getMaxResultWindow` is async, but it would be nice if we can flatten this out (`getMaxResultWindow` would become a data-request like most other I'd expect (?))
- introduce a new internal method on vector-layer that takes a source as a parameter. It would just be passed in by blended-layer (`super.syncDataSource(syncContext, source)`).

There's probably a few other ways, but since all our state needs to be managed in redux, I'm a little wary of introducing local state in our objects. I expect it will make troubleshooting/debugging harder and may hide subtle bugs.

Reply:
`_activeSource` and `_activeStyle` are not really mutable, it just toggles between `_document` and `_cluster`, so the underlying source is static. I could remove `active` from private state and just have getSource and getStyle methods return `document` or `cluster` source and track active static in a boolean or something.